### PR TITLE
Spesscliffs

### PR DESCRIPTION
**Pull request by RemieRichards**
:cl:
add: Added a system for turfs to render "cliffs", this really adds DEPTH to the game.
/:cl:
## Changes:
- Adds a system that lets turfs render "cliffs", another icon that gives the illusion of visual depth, it's a pretty neat looking effect.
- Turf types can have unique appearances for their "cliffs", for example gold mineral walls having gold mineral cliffs.
- The system uses a static list to reuse the same image per icon state, 1 icon state = 1 image for all turfs using that cliif.
- This is based on concepts by Bay12's @comma, although all code is written from scratch and redesigned by myself (static lists and reused images for example) The icon also comes from comma too (https://github.com/Baystation12/Baystation12/pull/13468), thanks comma!
## Images:

http://imgur.com/a/oibcH

### CODE REVIEW

**Comment by KorPhaeron** [comment]
Now add parallax

**Comment by GunHog** [comment]
Yeeeeeeeeeeeeeeeeeeesh

File: code/game/turfs/turf.dm [hunk 1]
```diff
@@ -320,3 +332,21 @@
 	SSair.remove_from_active(T0)
 	T0.CalculateAdjacentTurfs()
 	SSair.add_to_active(T0,1)
+
+
+/turf/proc/update_cliff()
+	if(cliff_images) //Not cut_overlays() because turf smoothing gets royally fugged.
+		var/list/L = list()
+		for(var/k in cliff_images)
```
Comment (RemieRichards) [inline code/game/turfs/turf.dm]: this loop saddens me, but I've made a feature request for the language to allow us to get lists of values/keys from Associative lists, if that's added than this whole block will just be ```overlays -= cliff_images.values```

**Comment by AnturK** [comment]
So these don't update at all with changes on neighbours ?

**Comment by RemieRichards** [comment]
@AnturK it's a NORTH/SOUTH effect, and it does indeed change on NORTH/SOUTH neighbour changes.

**Comment by WJohn** [comment]
So does this block movement in any way or is it purely a visual change?
Is it possible to change how big these cliff turfs are? like 2 tiles long or whatever?
Are they layer 2? Or do they cover things under them if something tries to be on the same tile as the cliff sprite?
Is it possible to smoothly stack multiple cliffs on top of eachother?
What about actual smoothing systems like walls/tables?

**Comment by RemieRichards** [comment]
@WJohn 
* it's below the turf visually, so no it doesn't block anything
* yes, if you gave it a say, 32x64 size dmi it would be effectively 2 turfs tall (would probably need some offsets, but doable)
* they are the layer of the turf, they are part of the turf's overlays
* there is only support for one per turf atm
* smoothing this would be overkill, I feel.

**Comment by WJohn** [comment]
Ah, well it's still definitely something. While we may not get huge plateaus sticking out of the ground, we can at least have pyramid structures that you can scale, which is fine by me.

**Comment by Niknakflak** [comment]
Still looks weird to me

**Comment by Jalleo** [comment]
Needs to have the shade start higher and be more gradual its too sharp for me.

**Comment by WJohn** [comment]
Those are just example pictures.

**Comment by Fox-McCloud** [comment]
>Now add parallax 

If you want the server to lag to hell and back and clients to barely move, then yeah.

Parallax is really neat and is very visually appealing, but the tax on the server is very high and the client performance issues it causes are crazy--even for those with super beefy PCs.

**Comment by MMMiracles** [comment]
The cliffs under windows looks off to me, but other than that, it looks pretty decent. Maybe a better cliff sprite for walls and it'd be good.

**Comment by RemieRichards** [comment]
@Fox-McCloud why'd you downvote? what are your reasons?

**Comment by Fox-McCloud** [comment]
Just not a fan of the way it looks, currently--I'll probably change my mind, later, if the sprites look a bit better. There's a few instances where it looks awkward---less like something that has depth and more something that just transitions from one square turf to another.

Take the singularity picture, for instance? That looks great, IMO. Where it looks really strange/wonky like in the area is here, for instance: http://i.imgur.com/ANrFWBb.png

That looks really awkward.

If that awkwardness is cleared up, then this will probably look a lot better/nicer, and I will gladly change my vote.

**Comment by RemieRichards** [comment]
@Fox-McCloud interesting you pick that one out as an odd one, as it's on of my favourites :S 
I only really asked for your opinion because you'd commented, downvoted, but not really stated why.

**Comment by KorPhaeron** [comment]
Mixing perspectives for turfs is probably gonna take some getting used to

**Comment by RemieRichards** [comment]
Yeah that's a common stance people have had so far in the bus "It's weird, but only because we're not used to it"

**Comment by PKPenguin321** [comment]
holy fuck that's the coolest thing ever

**Comment by optimumtact** [comment]
It doesn't look good because the wall is a flat top down sprite and then there is a strange floor tile beneath it.

I just don't see how this will ever work.

File: code/game/turfs/turf.dm (hunk 1 above)
Comment (xxalpha) [inline code/game/turfs/turf.dm]: Lay off the python.

Comment (RemieRichards) [inline code/game/turfs/turf.dm]: huh?

**Comment by xxalpha** [comment]
The idea is good, but this won't look good until the walls smooth with their cliffing because look at it.

**Comment by Core0verload** [comment]
Perspective in SS13 is already broken, and this breaks it even further.

**Comment by Alek2ander** [comment]
Lattices looks wonky with these. It looks like they should get an offset so they don't appear to be attavhed to the top of cliffs because that doesn't make sense.

**Comment by lzimann** [comment]
why remake when we can go 3D

**Comment by SkowronX** [comment]
A reason to sprite again perhaps?

**Comment by Bawhoppen** [comment]
It just doesn't look good, if anything we need it on top

**Comment by PJB3005** [comment]
@Fox-McCloud

> If you want the server to lag to hell and back and clients to barely move, then yeah.

Sound like somebody fucked up his port of parallax.

File: code/game/turfs/turf.dm (hunk 1 above)
Comment (PJB3005) [inline code/game/turfs/turf.dm]: @xxalpha this is not just a Python thing.

[.NET's `IDictionary<TKey, TValue>` defines a `Values` property too.](https://msdn.microsoft.com/en-us/library/cc673034(v=vs.110).aspx)

**Comment by gbasood** [comment]
/vg/ parallax works with basically no footprint.

**Comment by ike709** [comment]
I like the feature and the idea itself alot, but I don't really care much for how it actually looks... Maybe it could use a different sprite or something, I dunno.

File: code/game/turfs/turf.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/turfs/turf.dm]: @PJB3005 that's extra relevant, as I said in the bus (but failed to crosspost here) it was actually C# (.NET) Dictionaries that made me feel sad about this loop (and thus make the Feature Request, which lummox appears to support!)

**Comment by RemieRichards** [comment]
@ike709 art is most likely temporary, a sprite minion will no doubt redo cliffs.

**Comment by Kromgar** [comment]
image

It looks so weird after an explosion

**Comment by SkowronX** [comment]
It's because you're using half the default wall sprite.

**Comment by Kromgar** [comment]
image


On shuttles oh my

**Comment by KorPhaeron** [comment]
yuck

**Comment by Bawhoppen** [comment]
tear down this PR

**Comment by Kromgar** [comment]
Yeah this just really fucks with perspective. I'd be cool with cliffs on lavaland though. Like actual cliffs

**Comment by Kromgar** [comment]
image

**Comment by KorPhaeron** [comment]
shuttle stuff needs fixing

**Comment by Bawhoppen** [comment]
to be fair this does look cool
ss 2016-08-03 at 09 22 23

**Comment by Core0verload** [comment]
nope, still breaks my eyes

**Comment by PKPenguin321** [comment]
i saw this in game and it looked really really cool. shuttle needs to be fixed though.

one part that looked super cool was the north part of the bridge on metastation. looks absolutely badass with this

**Comment by lordpidey** [comment]
How does this interact with spacecarpet?  I could see this interfering with #19686

**Comment by Shadowlight213** [comment]
Meh. It just looks kinda dumb and really bad with the existing perspective.
It also makes small explosions breaching the hull make absolutely no sense, as with this change it looks like there's a huge chunk of thick metal below the floors, instead of a paper thin sheet like currently.
So, the price of floors for rcds, the code of manual building, as well as the floor explosion resistance would need to be increased in order to make this visual thing make sense.

**Comment by RemieRichards** [comment]
Shuttle stuff is because shuttles do matrix rotations, probably better to just 100% disable cliffing for shuttle turfs (I already did it for shuttle walls before putting the PR up, you can probably guess how ugly /that/ was)

@Shadowlight213 perhaps small explosions shouldn't breach then? the concept that the floor IS paper thin is ludicrous, even on an NT built space station.

A lot of talk comes from how it looks atm, so I'm tempted to either wait for a better sprite or disable the system and get it merged while waiting for a better sprite.
Remember, it doesn't -have- to look like this, the sprite can be redone and shuttle turfs can be set to no cliffs.

**Comment by GunHog** [comment]
Cliffs are quite amazing, and adds a nice bit of depth. I feel that /tg/ has been behind in graphical systems (no colored lighting, no smoothing lighting, no parallax, general playerbase aversion to graphical progress), yet one of the best for UIs.

I really do hope this makes it in. No big changes are going to be perfect. We can work out the bugs and make this a fantastic addition!

**Comment by scrdest** [comment]
This looks wrong on walls, because comma's original PR was designed for floors, applying it to walls makes no sense for the perspective. 

A floor on top of the cliff does; you get to see the edge where the two faces of the fake cube touch. A wall doesn't, because the sprite shows a cross-section of the wall from the top, with the front face of the actual non-cliff-effect wall missing and fucking up your 3D perception.

@Bawhoppen's image looks like those are flat platforms with a large metal tile on top, for instance.

**Comment by xxalpha** [comment]
>existing perspective

Is that a joke

**Comment by xxalpha** [comment]
Like, at least come up with real arguments if you don't like this. The only problem here is the cliff sprite being coder art.

**Comment by Kromgar** [comment]
We do have a perspctive alpha its top down its that simple

**Comment by xxalpha** [comment]
No it is not top down.

**Comment by PJB3005** [comment]
There is literally no unified perspective in SS13.

Walls and windows are top down.

A lot of things like vending machines for example are some sort of weird orthographic perspective (dimetric?)

Things like humans are 2.5d.

**Comment by Bawhoppen** [comment]
it just doesn't make sense, there may be no defined perspective, but the fact that the floors drop off so much underneath the floor itself is insane,

**Comment by xxalpha** [comment]
The cliff sprite needs to be good before it looks proper, but the idea itself is neat. I already see that other things want cliffing, like grilles, each cliff being of variable heights.

**Comment by WJohn** [comment]
I'd personally say start with this feature disabled on floors/walls and all that, but merge the code. I'd absolutely love to use this on lavaland for nice hills and mountains.

**Comment by YakumoChen** [comment]
looks ugly

**Comment by optimumtact** [comment]
Basically what @WJohn said

I'm not opposed to the system that gives mappers flexibility and may lead to a perspective resprite of our walls, but I don't think it should be on by default.

**Comment by PKPenguin321** [comment]
:(

**Comment by GunHog** [comment]
:( :(
Once again, /tg/ stays in the graphics past. No improvements to our antiquated lighting system, no parallax, and now no space cliffs. This attitude is leaving us in the dust as one of the worst major codebases for lighting and graphic effects.

**Comment by WJohn** [comment]
I mean we have pretty gradual lighting so there's that.

**Comment by Core0verload** [comment]
Once again, /tg/ stays in the graphics past, and that's good. We don't need SFX that look like shit.

**Comment by lzimann** [comment]
I agree with core

**Comment by GunHog** [comment]
/vg/'s (and Goon) smoothing and colored lighting looks substantially superior to ours in my opinion. So does parallax. When I look at their games and I look back at hours, it makes me feel as if we are trapped in cobwebs and shadows, an old relic that cannot ever be improved. This is depressing.

**Comment by xxalpha** [comment]
How does their smoothing look better exactly?

**Comment by Bawhoppen** [comment]
We need to port Tacolighting, by far better than any other.

**Comment by GunHog** [comment]
@Bawhoppen example?

**Comment by Bawhoppen** [comment]
I don't have the code, or an example, but it was really nice and subtle, as opposed to the extreme saturation of other colored lighting such as what PJB posted

**Comment by GunHog** [comment]
Here is /tg/ for comparison:
Very blocky lighting with obvious gradients based on tiles, with no coloration.
image

The bridge remains brightly lit due to having many more computers.

**Comment by lordpidey** [comment]
@Bawhoppen is the color saturation the problem of the lighting engine, or simply using lights that are too colored?

**Comment by Bawhoppen** [comment]
Probably the latter, but I can't be sure.

**Comment by PJB3005** [comment]
I will admit that the strength of the light emitted by most machines is extreme, but it looks cool IMO so eh.

**Comment by optimumtact** [comment]
Yeah I think rule of cool forgives the fact that the vg screens must give all the ingame characters some kind of eye strain

**Comment by Fox-McCloud** [comment]
@GunHog 

That lighting is also very expensive. We run Baylight, which is based upon TG's currently used lighting system, but it has an number of tweaks to it (some for the best, like not having to add/subtract lighting when picking up/dropping an object) including colored light support....that said, it's more expensive than TG's system.

Goon-lights is roughly x2.2-2.5 more expensive than that system.

**Comment by PJB3005** [comment]
> Baylight, which is based upon TG's currently used lighting system

No it isn't.

Baylights was made completely from scratch by Mloc and then also heavily optimized by me.

Even if you don't use coloured lighting it's superior in a multitude of ways. 

It's also faster than DAL but TG's object based DAL might be optimized since, not sure and I'm not gonna figure out because the code's a hot pile of shit.

> Goon-lights is roughly x2.2-2.5 more expensive than that system.

True, but while that sounds like a lot it's still not a performance problem by any stretch of the imagination.

**Pull request closed without merging**